### ISSUE TITLE
support for complex32 and complex64 as primitive types for zero-copy interop with numpy

### ISSUE THREAD

**Issue by maupardh1** [apache/arrow#39753]
### Describe the enhancement requested

Hello Arrow team,

First, I love Arrow - thank you so much for making this great project.

I am manipulating multi-dimensional array data (time-series like) that is produced from sensors as numpy arrays of type complex64. I would like to manipulate them in arrow for recording (feather and/or parquet formats) and distributed computing in the future (cu-df, dask, spark - most likely frameworks on top of arrow, but also cupy/scipy). This would also allow me to write column names and other schema metadata. I think it could be superior (and faster) than manipulating numpy arrays. 

It would be great to just call pa.array(np.array([1+ 2*1j,3+4*1j], dtype=np.complex64), type=pa.complex64()) but that type doesn't exist in Arrow. I haven't found a way to zero-copy a complex64 numpy array into a pyarrow array (my understanding is that only primitive types support zero-copy between arrow and numpy, and pa.binary(8) or struct attempts on top of numpy views so far have resulted in copies). I would also need to read it back from a feather/parquet format and potentially convert it to a numpy array if needed, and land back on np.complex64.

I think this has come up one or twice already: I found this thread: https://www.mail-archive.com/dev@arrow.apache.org/msg23352.html and this PR: https://github.com/apache/arrow/pull/10452 and thought I would also +1 this request, just in case.

If no first class support, do you see an alternative way to get zero-copy behavior? 

Thanks!
 

### Component(s)

Python

**Comment by zeroshade** [comment]
My initial thought here is that we could represent it via an extension type, are `complex32` and `complex64` internally represented using 32-bit and 64-bit values respectively? Or are they larger than 4 and 8 byte values?

@lidavidm @pitrou @bkietz thoughts?

**Comment by jorisvandenbossche** [comment]
General issue about complex support:

* https://github.com/apache/arrow/issues/16264

**Comment by pitrou** [comment]
This indeed could be a canonical extension type.

**Comment by maupardh1** [comment]
complex64 is 8 bytes in numpy  - the dtype description is ```<c8```.  Would a canonical extension type provide zero-copy interop?

**Comment by zeroshade** [comment]
Yes, a canonical extension type would provide zero-copy interop. The underlying representation could still be identical (a contiguous buffer of 8-byte values for example) which would have the underlying physical type of uint64 or something. Anything that recognizes the metadata and chooses to identify the extension type could then be able to interpret the values accordingly as complex64 values. I highly recommend putting together a proposal for these as canonical extension types.

**Comment by maupardh1** [comment]
got it - could the underlying type be binary(8) instead of uint64? feels a bit weird to represent a complex64 (2 32-bit floats) as a 64bit int? even if the underlying type representation doesn't end up getting used?

**Comment by zeroshade** [comment]
The only way for it to be zero-copy is if the underlying physical type matches the numpy representation. A struct of `{real: float32, imag: float32}` wouldn't be represented identically (it would be two separate buffers of float32s). A `fixedList` of floats makes sense and would provide the ability for zero-copy though. So that could work. 

Personally, I think that representing as an extension type with fixed-size-list would be easier and better than creating entirely new primitive types. But that's just my thought, anyone else wanna chime in here?

**Comment by pitrou** [comment]
Yes, I think `fixed_size_list<float, 2>` is a good start.

**Comment by maupardh1** [comment]
got it. Is there an example of an extension type PR/proposal that I could get inspiration from?

**Comment by bkietz** [comment]
FixedShapeTensor is a canonical extension type which [also uses `fixed_size_list` in its storage type](https://github.com/apache/arrow/blob/8bff1000e3c093a4b162c17cc04ca03949bb9e4b/cpp/src/arrow/extension/fixed_shape_tensor.h#L57). It was added in #8510, and mailing list [discussion](https://lists.apache.org/thread/oxvx0z0no2yyqsffzdc6nyjh6j4o6krs) and [vote](https://lists.apache.org/thread/f303yscj8gq4oggvvcbvf5zpo972l1kk) is archived.

As for numpy zero copy: we don't currently have a canonical extension type which maps to a numpy type like this one would. In order to support this, it'd be necessary to modify functions like [NumPyDtypeToArrow](https://github.com/apache/arrow/blob/8bff1000e3c093a4b162c17cc04ca03949bb9e4b/python/pyarrow/src/arrow/python/numpy_convert.cc#L133) which maps from a numpy type to an arrow DataType. That might be sufficient for `NumPyConverter::ConvertData` to [recognize zero-copy compatibility](https://github.com/apache/arrow/blob/8bff1000e3c093a4b162c17cc04ca03949bb9e4b/python/pyarrow/src/arrow/python/numpy_to_arrow.cc#L465-L470)

**Comment by williamfligor** [comment]
#notstale

**Comment by thisisnic** [comment]
This issue has been closed because there has been no activity for the 14 days since the stale warning. Feel free to reopen if this has been closed in error.
